accepted/2020/better-obsoletion/securestring-obsoletion.md

```diff
@@ -0,0 +1,221 @@
+
+The API surface clearly delineates buffers for which the caller maintains responsibility (the `ctor` and `CopyTo` methods). If the caller wishes that the plaintext data is never copied during a GC, then the destination buffer they pass to `CopyTo` should be allocated from the pinned object heap, the stack, or the native memory pool. The API surface gives them full control over that process.
+
+```cs
+// Example showing unshrouding a SensitiveData<char> into a temp char[].
+SensitiveData<char> sensitiveData = GetSensitiveStringData();
+char[] tempArray = GC.AllocateArray<char>(sensitiveData.Length, pinned: true);
+try
+{
+    sensitiveData.CopyTo(tempArray);
+    // use 'tempArray' here
+}
+finally
+{
+    Array.Clear(tempArray);
+}
+
+// Example showing unshrouding a SensitiveData<char> into a string.
+SensitiveData<char> sensitiveData = GetSensitiveStringData();
+string plaintextString = string.Create(
+    sensitiveData.Length, sensitiveData, (span, sensitiveData) =>
+    {
+        sensitiveData.CopyTo(span);
+    });
+```
+
+We would clearly document that this type is not intended to offer cryptographic protection of the buffer contents; it's simply intended to enforce more hygienic usage of data by callers.
+
+The type also intentionally does not give callers access to the raw backing span. This is because how the backing data is stored is strictly an implementation detail (it might not even be in-proc!), and exposing the span would leak that detail into the public API surface. The `CopyTo` method is the only way to get the raw contents. This also means that such buffers are immutable once created.
+
+__We should take care not to position this "sensitive buffer" type as a security feature for the purpose of achieving compliance.__ It's intended solely for code hygiene purposes and to avoid accidental disclosure. Scenarios like achieving PCI compliance require more fundamental runtime work, such as that under consideration at https://github.com/dotnet/runtime/issues/10480.
+
+This proposal intentionally excludes `ToString`, `ToArray`, or other methods which might unshroud the buffer. `ToString` is called frequently by library code as part of logging, and we always want unshrouding to be a _deliberate_ action. Additionally, we want the caller to maintain full control over the unshrouded buffer where the contents are written, which precludes "opinionated" APIs like `ToArray` which create such buffers themselves.
```

Comment:
I buy that `ToString()` is called by logging and other code paths that don't need to unshroud the buffer; but to me this doesn't mean we can (or shouldn't) provide a method like `UnshroudToString()` or `UnshroudToArray()`. Otherwise it will be hard to insert the type into existing code bases where virtually everyone uses strings today. The benefit of using a well-known method like `UnshroudToString()` is that it can be audited. If everyone just does the `string.Create` trick it's harder to audit (and harder to tool with an analyzer).

Reply:
hmmm I can see both sides of that argument :)

Reply:
Providing a well-known method would be fine in spirit IMO. The point of this type is to prevent *accidental* materialization of the string / array. If somebody is *intentionally* doing it, who are we to stop them?

Though to be honest the reason I didn't add such methods in the first place is that I assumed people would be nitpicky about how they wanted the information to be materialized. "I want it in a pinned array!" or whatever. `CopyTo` is intended to be a compromise which allows the caller to allocate the buffer themselves, then the secret is copied into that buffer. So `CopyTo` becomes the auditable method call. But if we need a simple accelerator, we could add one.

Reply:
I'd much rather not. A single CopyTo is auditable, adding more methods means more noise, and more chance for a bad audit to miss things. Keeping it to a single method is better in terms of auditing, there's less things to miss.